Does not build with `extended` on

I have `extended = true` set in `config.toml`, but this seems to be causing the build to fail (master HEAD at time of writing), where it wasn't before.

```
$ ./x.py -i build
Updating only changed submodules
Updating submodule src/tools/clippy
Submodule path 'src/tools/clippy': checked out 'c5b39a5917ffc0f1349b6e414fa3b874fdcf8429'
Submodules updated in 1.15 seconds
    Finished dev [unoptimized] target(s) in 0.0 secs
Building stage0 std artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling core v0.0.0 (file:///Users/alex/Software/rust/src/libcore)
   Compiling unwind v0.0.0 (file:///Users/alex/Software/rust/src/libunwind)
   Compiling build_helper v0.1.0 (file:///Users/alex/Software/rust/src/build_helper)
   Compiling compiler_builtins v0.0.0 (file:///Users/alex/Software/rust/src/rustc/compiler_builtins_shim)
   Compiling rustc_tsan v0.0.0 (file:///Users/alex/Software/rust/src/librustc_tsan)
   Compiling rustc_asan v0.0.0 (file:///Users/alex/Software/rust/src/librustc_asan)
   Compiling alloc_jemalloc v0.0.0 (file:///Users/alex/Software/rust/src/liballoc_jemalloc)
   Compiling std v0.0.0 (file:///Users/alex/Software/rust/src/libstd)
   Compiling libc v0.0.0 (file:///Users/alex/Software/rust/src/rustc/libc_shim)
   Compiling alloc v0.0.0 (file:///Users/alex/Software/rust/src/liballoc)
   Compiling std_unicode v0.0.0 (file:///Users/alex/Software/rust/src/libstd_unicode)
   Compiling alloc_system v0.0.0 (file:///Users/alex/Software/rust/src/liballoc_system)
   Compiling panic_abort v0.0.0 (file:///Users/alex/Software/rust/src/libpanic_abort)
   Compiling panic_unwind v0.0.0 (file:///Users/alex/Software/rust/src/libpanic_unwind)
    Finished release [optimized] target(s) in 50.30 secs
Copying stage0 std from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Building stage0 test artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling getopts v0.2.17
   Compiling term v0.0.0 (file:///Users/alex/Software/rust/src/libterm)
   Compiling test v0.0.0 (file:///Users/alex/Software/rust/src/libtest)
    Finished release [optimized] target(s) in 8.28 secs
Copying stage0 test from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Building stage0 compiler artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling libc v0.2.40
   Compiling stable_deref_trait v1.0.0
   Compiling smallvec v0.6.0
   Compiling cfg-if v0.1.2
   Compiling bitflags v1.0.1
   Compiling serialize v0.0.0 (file:///Users/alex/Software/rust/src/libserialize)
   Compiling rustc_target v0.0.0 (file:///Users/alex/Software/rust/src/librustc_target)
   Compiling unicode-width v0.1.4
   Compiling scoped-tls v0.1.1
   Compiling termcolor v0.3.6
   Compiling syntax v0.0.0 (file:///Users/alex/Software/rust/src/libsyntax)
   Compiling rustc v0.0.0 (file:///Users/alex/Software/rust/src/librustc)
   Compiling remove_dir_all v0.5.1
   Compiling rustc-demangle v0.1.7
   Compiling byteorder v1.2.2
   Compiling graphviz v0.0.0 (file:///Users/alex/Software/rust/src/libgraphviz)
   Compiling lazy_static v1.0.0
   Compiling rustc-serialize v0.3.24
   Compiling rustc_metadata v0.0.0 (file:///Users/alex/Software/rust/src/librustc_metadata)
   Compiling fmt_macros v0.0.0 (file:///Users/alex/Software/rust/src/libfmt_macros)
   Compiling lazy_static v0.2.11
   Compiling rustc_incremental v0.0.0 (file:///Users/alex/Software/rust/src/librustc_incremental)
   Compiling quick-error v1.2.1
   Compiling rustc_platform_intrinsics v0.0.0 (file:///Users/alex/Software/rust/src/librustc_platform_intrinsics)
   Compiling ar v0.3.1
   Compiling rustc_driver v0.0.0 (file:///Users/alex/Software/rust/src/librustc_driver)
   Compiling log v0.4.1
   Compiling owning_ref v0.3.3
   Compiling rand v0.4.2
   Compiling atty v0.2.8
   Compiling log_settings v0.1.1
   Compiling humantime v1.1.1
   Compiling backtrace v0.3.6
   Compiling miniz-sys v0.1.10
   Compiling ena v0.9.2
   Compiling rustc_cratesio_shim v0.0.0 (file:///Users/alex/Software/rust/src/librustc_cratesio_shim)
   Compiling jobserver v0.1.11
   Compiling env_logger v0.5.8
   Compiling rustc_apfloat v0.0.0 (file:///Users/alex/Software/rust/src/librustc_apfloat)
   Compiling parking_lot_core v0.2.14
   Compiling tempdir v0.3.7
   Compiling flate2 v1.0.1
   Compiling parking_lot v0.5.5
   Compiling rustc_data_structures v0.0.0 (file:///Users/alex/Software/rust/src/librustc_data_structures)
   Compiling rls-span v0.4.0
   Compiling rls-data v0.15.0
   Compiling syntax_pos v0.0.0 (file:///Users/alex/Software/rust/src/libsyntax_pos)
   Compiling arena v0.0.0 (file:///Users/alex/Software/rust/src/libarena)
   Compiling rustc_errors v0.0.0 (file:///Users/alex/Software/rust/src/librustc_errors)
   Compiling proc_macro v0.0.0 (file:///Users/alex/Software/rust/src/libproc_macro)
   Compiling rustc_const_math v0.0.0 (file:///Users/alex/Software/rust/src/librustc_const_math)
   Compiling syntax_ext v0.0.0 (file:///Users/alex/Software/rust/src/libsyntax_ext)
   Compiling rustc_typeck v0.0.0 (file:///Users/alex/Software/rust/src/librustc_typeck)
   Compiling rustc_mir v0.0.0 (file:///Users/alex/Software/rust/src/librustc_mir)
   Compiling rustc_allocator v0.0.0 (file:///Users/alex/Software/rust/src/librustc_allocator)
   Compiling rustc_traits v0.0.0 (file:///Users/alex/Software/rust/src/librustc_traits)
   Compiling rustc_resolve v0.0.0 (file:///Users/alex/Software/rust/src/librustc_resolve)
   Compiling rustc_plugin v0.0.0 (file:///Users/alex/Software/rust/src/librustc_plugin)
   Compiling rustc_privacy v0.0.0 (file:///Users/alex/Software/rust/src/librustc_privacy)
   Compiling rustc_save_analysis v0.0.0 (file:///Users/alex/Software/rust/src/librustc_save_analysis)
   Compiling rustc_lint v0.0.0 (file:///Users/alex/Software/rust/src/librustc_lint)
   Compiling rustc_trans_utils v0.0.0 (file:///Users/alex/Software/rust/src/librustc_trans_utils)
   Compiling rustc_borrowck v0.0.0 (file:///Users/alex/Software/rust/src/librustc_borrowck)
   Compiling rustc_passes v0.0.0 (file:///Users/alex/Software/rust/src/librustc_passes)
   Compiling rustc-main v0.0.0 (file:///Users/alex/Software/rust/src/rustc)
    Finished release [optimized] target(s) in 694.18 secs
Copying stage0 rustc from stage0 (x86_64-apple-darwin -> x86_64-apple-darwin / x86_64-apple-darwin)
Building stage0 codegen artifacts (x86_64-apple-darwin -> x86_64-apple-darwin, llvm)
   Compiling build_helper v0.1.0 (file:///Users/alex/Software/rust/src/build_helper)
   Compiling rustc_trans v0.0.0 (file:///Users/alex/Software/rust/src/librustc_trans)
   Compiling cc v1.0.10
   Compiling num_cpus v1.8.0
   Compiling rustc_llvm v0.0.0 (file:///Users/alex/Software/rust/src/librustc_llvm)
    Finished release [optimized] target(s) in 67.70 secs
Assembling stage1 compiler (x86_64-apple-darwin)
Building stage1 std artifacts (x86_64-apple-darwin -> x86_64-apple-darwin)
   Compiling core v0.0.0 (file:///Users/alex/Software/rust/src/libcore)
   Compiling unwind v0.0.0 (file:///Users/alex/Software/rust/src/libunwind)
   Compiling compiler_builtins v0.0.0 (file:///Users/alex/Software/rust/src/rustc/compiler_builtins_shim)
   Compiling alloc_jemalloc v0.0.0 (file:///Users/alex/Software/rust/src/liballoc_jemalloc)
   Compiling rustc_asan v0.0.0 (file:///Users/alex/Software/rust/src/librustc_asan)
   Compiling rustc_tsan v0.0.0 (file:///Users/alex/Software/rust/src/librustc_tsan)
   Compiling std v0.0.0 (file:///Users/alex/Software/rust/src/libstd)
error: internal compiler error: librustc/ty/context.rs:275: node unknown node (id=1) with HirId::owner DefId(0/0:0 ~ core[7c27]) cannot be placed in TypeckTables with local_id_root DefId(0/0:1840 ~ core[7c27]::panicking[0]::panic_fmt[0])

thread 'main' panicked at 'Box<Any>', librustc_errors/lib.rs:546:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: aborting due to previous error


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.27.0-dev running on x86_64-apple-darwin

note: compiler flags: -Z save-analysis -Z osx-rpath-install-name -Z force-unstable-if-unmarked -C opt-level=2 -C prefer-dynamic -C debug-assertions=y -C codegen-units=8 -C link-args=-Wl,-rpath,@loader_path/../lib --crate-type lib

note: some of the compiler flags provided by cargo are hidden

error: Could not compile `core`.

Caused by:
  process didn't exit successfully: `/Users/alex/Software/rust/build/bootstrap/debug/rustc --crate-name core libcore/lib.rs --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=be5db4a71f87466b -C extra-filename=-be5db4a71f87466b --out-dir /Users/alex/Software/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps --target x86_64-apple-darwin -L dependency=/Users/alex/Software/rust/build/x86_64-apple-darwin/stage1-std/x86_64-apple-darwin/release/deps -L dependency=/Users/alex/Software/rust/build/x86_64-apple-darwin/stage1-std/release/deps` (exit code: 101)
command did not execute successfully: "/Users/alex/Software/rust/build/x86_64-apple-darwin/stage0/bin/cargo" "build" "--target" "x86_64-apple-darwin" "-j" "8" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/Users/alex/Software/rust/src/libstd/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101
thread 'main' panicked at 'cargo must succeed', bootstrap/compile.rs:1091:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failed to run: /Users/alex/Software/rust/build/bootstrap/debug/bootstrap -i build
Build completed unsuccessfully in 0:15:41
```

Is this a bug in the build process?

I'm on macOS 10.13.4, incidentally.

@alexcrichton Yeah... I turned off `debug-assertions` and `debuginfo` and rustc builds now. The `miri` tool fails to build, however.

```
Building stage2 tool miri (x86_64-apple-darwin)
   Compiling miri v0.1.0 (file:///Users/alex/Software/rust/src/tools/miri)
   Compiling byteorder v1.2.2
error[E0432]: unresolved import `syntax::abi`
 --> tools/miri/miri/fn_call.rs:7:13
  |
7 | use syntax::abi::Abi;
  |             ^^^ Could not find `abi` in `syntax`

error[E0308]: match arms have incompatible types
   --> tools/miri/miri/fn_call.rs:178:25
    |
178 |           let link_name = match attr::first_attr_value_str_by_name(&attrs, "link_name") {
    |  _________________________^
179 | |             Some(name) => name.as_str(),
180 | |             None => self.tcx.item_name(def_id),
    | |                     -------------------------- match arm with an incompatible type
181 | |         };
    | |_________^ expected struct `syntax::symbol::LocalInternedString`, found struct `syntax::symbol::InternedString`
    |
    = note: expected type `syntax::symbol::LocalInternedString`
               found type `syntax::symbol::InternedString`

error[E0608]: cannot index into a value of type `syntax::symbol::InternedString`
  --> tools/miri/miri/intrinsic.rs:32:31
   |
32 |         let intrinsic_name = &self.tcx.item_name(instance.def_id())[..];
   |                               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `to_ty` found for type `rustc::ty::layout::Primitive` in the current scope
   --> tools/miri/miri/validation.rs:461:47
    |
461 |                         return Ok(discr.value.to_ty(tcx))
    |                                               ^^^^^
    |
    = help: items from traits can only be used if the trait is in scope
    = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
            candidate #1: `use rustc::ty::layout::PrimitiveExt;`

error: aborting due to 4 previous errors

Some errors occurred: E0308, E0432, E0599, E0608.
For more information about an error, try `rustc --explain E0308`.
error: Could not compile `miri`.
```

--- a/src/librustc/ty/context.rs
+++ b/src/librustc/ty/context.rs
@@ -205,26 +205,24 @@ pub struct LocalTableInContext<'a, V> {
 fn validate_hir_id_for_typeck_tables(local_id_root: Option<DefId>,
                                      hir_id: hir::HirId,
                                      mut_access: bool) {
-    if cfg!(debug_assertions) {
-        if let Some(local_id_root) = local_id_root {
-            if hir_id.owner != local_id_root.index {
-                ty::tls::with(|tcx| {
-                    bug!("node {} with HirId::owner {:?} cannot be placed in \
-                          TypeckTables with local_id_root {:?}",
-                         tcx.hir().node_to_string(hir_id),
-                         DefId::local(hir_id.owner),
-                         local_id_root)
-                });
-            }
-        } else {
-            // We use "Null Object" TypeckTables in some of the analysis passes.
-            // These are just expected to be empty and their `local_id_root` is
-            // `None`. Therefore we cannot verify whether a given `HirId` would
-            // be a valid key for the given table. Instead we make sure that
-            // nobody tries to write to such a Null Object table.
-            if mut_access {
-                bug!("access to invalid TypeckTables")
-            }
+    if let Some(local_id_root) = local_id_root {
+        if hir_id.owner != local_id_root.index {
+            ty::tls::with(|tcx| {
+                bug!("node {} with HirId::owner {:?} cannot be placed in \
+                        TypeckTables with local_id_root {:?}",
+                        tcx.hir().node_to_string(hir_id),
+                        DefId::local(hir_id.owner),
+                        local_id_root)
+            });
+        }
+    } else {
+        // We use "Null Object" TypeckTables in some of the analysis passes.
+        // These are just expected to be empty and their `local_id_root` is
+        // `None`. Therefore we cannot verify whether a given `HirId` would
+        // be a valid key for the given table. Instead we make sure that
+        // nobody tries to write to such a Null Object table.
+        if mut_access {
+            bug!("access to invalid TypeckTables")
         }
     }
 }

--- a/src/librustc_save_analysis/dump_visitor.rs
+++ b/src/librustc_save_analysis/dump_visitor.rs
@@ -283,36 +283,32 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
     ) {
         debug!("process_method: {}:{}", id, ident);
 
-        if let Some(mut method_data) = self.save_ctxt.get_method_data(id, ident, span) {
-            let sig_str = crate::make_signature(&sig.decl, &generics);
-            if body.is_some() {
-                self.nest_tables(
-                    id,
-                    |v| v.process_formals(&sig.decl.inputs, &method_data.qualname),
-                );
+        let hir_id = self.tcx.hir().node_to_hir_id(id);
+        self.nest_tables(id, |v| {
+            if let Some(mut method_data) = v.save_ctxt.get_method_data(id, ident, span) {
+                v.process_formals(&sig.decl.inputs, &method_data.qualname);
+                v.process_generic_params(&generics, &method_data.qualname, id);
+
+                method_data.value = crate::make_signature(&sig.decl, &generics);
+                method_data.sig = sig::method_signature(id, ident, generics, sig, &v.save_ctxt);
+
+                v.dumper.dump_def(&access_from_vis!(v.save_ctxt, vis, hir_id), method_data);
             }
 
-            self.process_generic_params(&generics, &method_data.qualname, id);
+            // walk arg and return types
+            for arg in &sig.decl.inputs {
+                v.visit_ty(&arg.ty);
+            }
 
-            method_data.value = sig_str;
-            method_data.sig = sig::method_signature(id, ident, generics, sig, &self.save_ctxt);
-            let hir_id = self.tcx.hir().node_to_hir_id(id);
-            self.dumper.dump_def(&access_from_vis!(self.save_ctxt, vis, hir_id), method_data);
-        }
+            if let ast::FunctionRetTy::Ty(ref ret_ty) = sig.decl.output {
+                v.visit_ty(ret_ty);
+            }
 
-        // walk arg and return types
-        for arg in &sig.decl.inputs {
-            self.visit_ty(&arg.ty);
-        }
-
-        if let ast::FunctionRetTy::Ty(ref ret_ty) = sig.decl.output {
-            self.visit_ty(ret_ty);
-        }
-
-        // walk the fn body
-        if let Some(body) = body {
-            self.nest_tables(id, |v| v.visit_block(body));
-        }
+            // walk the fn body
+            if let Some(body) = body {
+                v.visit_block(body);
+            }
+        });
     }
 
     fn process_struct_field_def(&mut self, field: &ast::StructField, parent_id: NodeId) {
@@ -377,26 +373,26 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
         ty_params: &'l ast::Generics,
         body: &'l ast::Block,
     ) {
-        if let Some(fn_data) = self.save_ctxt.get_item_data(item) {
-            down_cast_data!(fn_data, DefData, item.span);
-            self.nest_tables(
-                item.id,
-                |v| v.process_formals(&decl.inputs, &fn_data.qualname),
-            );
-            self.process_generic_params(ty_params, &fn_data.qualname, item.id);
-            let hir_id = self.tcx.hir().node_to_hir_id(item.id);
-            self.dumper.dump_def(&access_from!(self.save_ctxt, item, hir_id), fn_data);
-        }
+        let hir_id = self.tcx.hir().node_to_hir_id(item.id);
+        self.nest_tables(item.id, |v| {
+            if let Some(fn_data) = v.save_ctxt.get_item_data(item) {
+                down_cast_data!(fn_data, DefData, item.span);
+                v.process_formals(&decl.inputs, &fn_data.qualname);
+                v.process_generic_params(ty_params, &fn_data.qualname, item.id);
 
-        for arg in &decl.inputs {
-            self.visit_ty(&arg.ty);
-        }
+                v.dumper.dump_def(&access_from!(v.save_ctxt, item, hir_id), fn_data);
+            }
 
-        if let ast::FunctionRetTy::Ty(ref ret_ty) = decl.output {
-            self.visit_ty(&ret_ty);
-        }
+            for arg in &decl.inputs {
+                v.visit_ty(&arg.ty)
+            }
 
-        self.nest_tables(item.id, |v| v.visit_block(&body));
+            if let ast::FunctionRetTy::Ty(ref ret_ty) = decl.output {
+                v.visit_ty(&ret_ty);
+            }
+
+            v.visit_block(&body);
+        });
     }
 
     fn process_static_or_const_item(

--- a/src/librustc_save_analysis/dump_visitor.rs
+++ b/src/librustc_save_analysis/dump_visitor.rs
@@ -385,7 +385,12 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
             }
 
             if let ast::FunctionRetTy::Ty(ref ret_ty) = decl.output {
-                v.visit_ty(&ret_ty);
+                if let ast::TyKind::ImplTrait(..) = ret_ty.node {
+                    // FIXME: Opaque type desugaring prevents us from easily
+                    // processing trait bounds. See `visit_ty` for more details.
+                } else {
+                    v.visit_ty(&ret_ty);
+                }
             }
 
             v.visit_block(&body);
@@ -1438,6 +1443,18 @@ impl<'l, 'tcx> Visitor<'l> for DumpVisitor<'l, 'tcx> {
             ast::TyKind::Array(ref element, ref length) => {
                 self.visit_ty(element);
                 self.nest_tables(length.id, |v| v.visit_expr(&length.value));
+            }
+            ast::TyKind::ImplTrait(id, ref bounds) => {
+                // FIXME: As of writing, the opaque type lowering introduces
+                // another DefPath scope/segment (used to declare the resulting
+                // opaque type item).
+                // However, the synthetic scope does *not* have associated
+                // typeck tables, which means we can't nest it and we fire an
+                // assertion when resolving the qualified type paths in trait
+                // bounds...
+                // This will panic if called on return type `impl Trait`, which
+                // we guard against in `process_fn`.
+                self.nest_tables(id, |v| v.process_bounds(bounds));
             }
             _ => visit::walk_ty(self, t),
         }

--- a/src/librustc_save_analysis/dump_visitor.rs
+++ b/src/librustc_save_analysis/dump_visitor.rs
@@ -130,6 +130,10 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
         self.save_ctxt.span_from_span(span)
     }
 
+    fn lookup_def_id(&self, ref_id: NodeId) -> Option<DefId> {
+        self.save_ctxt.lookup_def_id(ref_id)
+    }
+
     pub fn dump_crate_info(&mut self, name: &str, krate: &ast::Crate) {
         let source_file = self.tcx.sess.local_crate_source_file.as_ref();
         let crate_root = source_file.map(|source_file| {
@@ -220,13 +224,6 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
             if let Some(data) = self.save_ctxt.get_path_segment_data(seg) {
                 self.dumper.dump_ref(data);
             }
-        }
-    }
-
-    fn lookup_def_id(&self, ref_id: NodeId) -> Option<DefId> {
-        match self.save_ctxt.get_path_res(ref_id) {
-            Res::PrimTy(..) | Res::SelfTy(..) | Res::Err => None,
-            def => Some(def.def_id()),
         }
     }
 

--- a/src/librustc_save_analysis/dump_visitor.rs
+++ b/src/librustc_save_analysis/dump_visitor.rs
@@ -1109,11 +1109,7 @@ impl<'l, 'tcx> DumpVisitor<'l, 'tcx> {
                 // FIXME: uses of the assoc type should ideally point to this
                 // 'def' and the name here should be a ref to the def in the
                 // trait.
-                for bound in bounds.iter() {
-                    if let ast::GenericBound::Trait(trait_ref, _) = bound {
-                        self.process_path(trait_ref.trait_ref.ref_id, &trait_ref.trait_ref.path)
-                    }
-                }
+                self.process_bounds(&bounds);
             }
             ast::ImplItemKind::Macro(_) => {}
         }

--- a/src/librustc_save_analysis/dump_visitor.rs
+++ b/src/librustc_save_analysis/dump_visitor.rs
@@ -1356,10 +1356,10 @@ impl<'l, 'tcx> Visitor<'l> for DumpVisitor<'l, 'tcx> {
                 self.visit_ty(&ty);
                 self.process_generic_params(ty_params, &qualname, item.id);
             }
-            OpaqueTy(ref _bounds, ref ty_params) => {
+            OpaqueTy(ref bounds, ref ty_params) => {
                 let qualname = format!("::{}",
                     self.tcx.def_path_str(self.tcx.hir().local_def_id_from_node_id(item.id)));
-                // FIXME do something with _bounds
+
                 let value = String::new();
                 if !self.span.filter_generated(item.ident.span) {
                     let span = self.span_from_span(item.ident.span);
@@ -1385,6 +1385,7 @@ impl<'l, 'tcx> Visitor<'l> for DumpVisitor<'l, 'tcx> {
                     );
                 }
 
+                self.process_bounds(bounds);
                 self.process_generic_params(ty_params, &qualname, item.id);
             }
             Mac(_) => (),

--- a/src/librustc_save_analysis/lib.rs
+++ b/src/librustc_save_analysis/lib.rs
@@ -312,7 +312,7 @@ impl<'l, 'tcx> SaveContext<'l, 'tcx> {
                     let impl_id = self.next_impl_id();
                     let span = self.span_from_span(sub_span);
 
-                    let type_data = self.lookup_ref_id(typ.id);
+                    let type_data = self.lookup_def_id(typ.id);
                     type_data.map(|type_data| {
                         Data::RelationData(Relation {
                             kind: RelationKind::Impl {
@@ -322,7 +322,7 @@ impl<'l, 'tcx> SaveContext<'l, 'tcx> {
                             from: id_from_def_id(type_data),
                             to: trait_ref
                                 .as_ref()
-                                .and_then(|t| self.lookup_ref_id(t.ref_id))
+                                .and_then(|t| self.lookup_def_id(t.ref_id))
                                 .map(id_from_def_id)
                                 .unwrap_or_else(|| null_id()),
                         },
@@ -495,7 +495,7 @@ impl<'l, 'tcx> SaveContext<'l, 'tcx> {
     }
 
     pub fn get_trait_ref_data(&self, trait_ref: &ast::TraitRef) -> Option<Ref> {
-        self.lookup_ref_id(trait_ref.ref_id).and_then(|def_id| {
+        self.lookup_def_id(trait_ref.ref_id).and_then(|def_id| {
             let span = trait_ref.path.span;
             if generated_code(span) {
                 return None;
@@ -870,7 +870,7 @@ impl<'l, 'tcx> SaveContext<'l, 'tcx> {
         })
     }
 
-    fn lookup_ref_id(&self, ref_id: NodeId) -> Option<DefId> {
+    fn lookup_def_id(&self, ref_id: NodeId) -> Option<DefId> {
         match self.get_path_res(ref_id) {
             Res::PrimTy(_) | Res::SelfTy(..) | Res::Err => None,
             def => Some(def.def_id()),

--- a/src/test/ui/save-analysis/issue-63663.rs
+++ b/src/test/ui/save-analysis/issue-63663.rs
@@ -1,15 +1,20 @@
 // check-pass
 // compile-flags: -Zsave-analysis
 
-// Check that this doesn't ICE when processing associated const in formal
-// argument and return type of functions defined inside function/method scope.
-
 pub trait Trait {
     type Assoc;
 }
 
 pub struct A;
 
+trait Generic<T> {}
+impl<T> Generic<T> for () {}
+
+// Don't ICE when resolving type paths in return type `impl Trait`
+fn assoc_in_opaque_type_bounds<U: Trait>() -> impl Generic<U::Assoc> {}
+
+// Check that this doesn't ICE when processing associated const in formal
+// argument and return type of functions defined inside function/method scope.
 pub fn func() {
     fn _inner1<U: Trait>(_: U::Assoc) {}
     fn _inner2<U: Trait>() -> U::Assoc { unimplemented!() }

--- a/src/test/ui/save-analysis/issue-63663.rs
+++ b/src/test/ui/save-analysis/issue-63663.rs
@@ -1,0 +1,23 @@
+// check-pass
+// compile-flags: -Zsave-analysis
+
+// Check that this doesn't ICE when processing associated const in formal
+// argument and return type of functions defined inside function/method scope.
+
+pub trait Trait {
+    type Assoc;
+}
+
+pub struct A;
+
+pub fn func() {
+    fn _inner1<U: Trait>(_: U::Assoc) {}
+    fn _inner2<U: Trait>() -> U::Assoc { unimplemented!() }
+
+    impl A {
+        fn _inner1<U: Trait>(self, _: U::Assoc) {}
+        fn _inner2<U: Trait>(self) -> U::Assoc { unimplemented!() }
+    }
+}
+
+fn main() {}

